src/main/java/com/exasol/spark/s3/ExasolBatchWrite.java

```diff
@@ -0,0 +1,164 @@
+package com.exasol.spark.s3;
+
+import static com.exasol.spark.s3.Constants.INTERMEDIATE_DATA_PATH;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.spark.sql.connector.write.*;
+
+import com.exasol.errorreporting.ExaError;
+
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Request;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Response;
+import software.amazon.awssdk.services.s3.model.S3Object;
+
+/**
+ * An Exasol {@link BatchWrite} implementation.
+ */
+public class ExasolBatchWrite implements BatchWrite {
+    private static final Logger LOGGER = Logger.getLogger(ExasolBatchWrite.class.getName());
+
+    private final ExasolOptions options;
+    private final BatchWrite delegate;
+
+    /**
+     * Creates a new instance of {@link ExasolBatchWrite}.
+     *
+     * @param options  user provided options
+     * @param delegate delegate {@code CSV} batch write
+     */
+    public ExasolBatchWrite(final ExasolOptions options, final Write delegate) {
+        this.options = options;
+        this.delegate = delegate.toBatch();
+    }
+
+    @Override
+    public DataWriterFactory createBatchWriterFactory(final PhysicalWriteInfo info) {
+        return delegate.createBatchWriterFactory(info);
+    }
+
+    @Override
+    public boolean useCommitCoordinator() {
+        return delegate.useCommitCoordinator();
+    }
+
+    @Override
+    public void abort(final WriterCommitMessage[] messages) {
+        LOGGER.info("Running abort stage of the job.");
+        cleanup();
+        delegate.abort(messages);
+    }
+
+    private void cleanup() {
+        LOGGER.info(() -> "Running cleanup process for directory '" + this.options.get(INTERMEDIATE_DATA_PATH) + "'.");
+    }
+
+    @Override
+    public void commit(final WriterCommitMessage[] messages) {
+        LOGGER.info("Committing the file writing stage of the job.");
+        delegate.commit(messages);
+        prepareIntermediateData();
+    }
+
+    private void prepareIntermediateData() {
+        final long start = System.currentTimeMillis();
+        final String table = this.options.getTable();
+        final String query = new S3ImportQueryGenerator(options).generateQuery();
+        try (final Connection connection = new ExasolConnectionFactory(this.options).getConnection();
+                final Statement stmt = connection.createStatement()) {
+            connection.setAutoCommit(false);
+            final int rows = stmt.executeUpdate(query);
+            connection.commit();
```

Comment:
Maybe extract to own function, might make this a bit less convoluted

Reply:
Done.